tests/unit/openstack_cloud/test_openstack_runner_manager.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for unit-testing OpenStack runner manager."""
+from datetime import datetime, timedelta
+from pathlib import Path
+from unittest.mock import MagicMock
+
+import pytest
+
+from github_runner_manager.metrics import runner
+from github_runner_manager.metrics.storage import MetricsStorage, StorageManager
+from github_runner_manager.openstack_cloud import openstack_runner_manager
+from github_runner_manager.openstack_cloud.openstack_runner_manager import (
+    OUTDATED_METRICS_STORAGE_IN_SECONDS,
+    OpenStackRunnerManager,
+)
+
+
+@pytest.mark.parametrize(
+    "healthy_runner_names, unhealthy_runner_names, undecided_runner_storage, "
+    "expected_storage_to_be_extracted",
+    [
+        pytest.param(
+            default_healthy := {"healthy1", "healthy2"},
+            default_unhealthy := {"unhealthy1", "unhealthy2"},
+            default_undecided := {
+                ("in_construction", datetime.now()),
+                (
+                    "dangling",
+                    datetime.now() - timedelta(seconds=OUTDATED_METRICS_STORAGE_IN_SECONDS + 1),
+                ),
+            },
+            default_result := default_unhealthy | {"dangling"},
+            id="one dangling",
+        ),
+        pytest.param(
+            default_healthy,
+            default_unhealthy,
+            {
+                ("in_construction", datetime.now()),
+                (
+                    "dangling",
+                    datetime.now() - timedelta(seconds=OUTDATED_METRICS_STORAGE_IN_SECONDS + 1),
+                ),
+                (
+                    "dangling2",
+                    datetime.now() - timedelta(seconds=OUTDATED_METRICS_STORAGE_IN_SECONDS + 1),
+                ),
+            },
+            default_unhealthy | {"dangling", "dangling2"},
+            id="two dangling",
+        ),
+        pytest.param(
+            default_healthy,
+            default_unhealthy,
+            {("in_construction", datetime.now())},
+            default_unhealthy,
+            id="no dangling",
+        ),
+        pytest.param(
+            default_healthy,
+            set(),
+            default_undecided,
+            {"dangling"},
+            id="no unhealthy",
+        ),
+        pytest.param(
+            default_healthy,
+            default_unhealthy,
+            set(),
+            default_unhealthy,
+            id="no undecided",
+        ),
+        pytest.param(
+            set(),
+            default_unhealthy,
+            default_undecided,
+            default_result,
+            id="no healthy",
+        ),
+    ],
+)
+def test__cleanup_extract_metrics(
```

Comment:
I am not a big fan of testing private implementations, but I think it would be too much of a design change to move this implementation elsewhere for now.